File: packages/dotnet/NaturalSpacing.Core/Grapheme17.cs
using System.Text;

namespace NaturalSpacing.Core;

internal static class Grapheme17
{
    internal static IReadOnlyList<int> Boundaries(string text)
    {
        if (text.Length == 0) return [0];

        var result = new List<int> { 0 };
        var first = Rune.GetRuneAt(text, 0);
        var cursor = first.Utf16SequenceLength;
        var categoryBefore = Grapheme17Data.Category(first.Value);
        var state = NextState(1, categoryBefore, first.Value);

        while (cursor < text.Length)
        {
            var rune = Rune.GetRuneAt(text, cursor);
            var categoryAfter = Grapheme17Data.Category(rune.Value);
            var pairMask = Grapheme17Data.PairMasks[(categoryBefore << 4) | categoryAfter];
            if ((state & pairMask) == 0) result.Add(cursor);
            state = NextState(state, categoryAfter, rune.Value);
            cursor += rune.Utf16SequenceLength;
            categoryBefore = categoryAfter;
        }
        result.Add(text.Length);
        return result;
    }

    private static int NextState(int state, int category, int codePoint) => category switch
    {
        3 => (state & 32) != 0 ? NextExtend(state, codePoint) : state & 21,
        4 => 17,
        10 => (state & 2) ^ 3,
        14 => ((state & 16) >> 2) | (state & 41),
        15 => 33,
        _ => 1,
    };

    private static int NextExtend(int state, int codePoint)
    {
        if (codePoint == 0x200C) return state & 21;
        return (state & 8) != 0 || Grapheme17Data.IsLinker(codePoint)
            ? (state & 21) | 40
            : (state & 21) | 32;
    }
}

File: packages/kotlin/src/main/kotlin/dev/naturalspacing/core/Grapheme17.kt
package dev.naturalspacing.core

internal object Grapheme17 {
    fun boundaries(text: String): IntArray {
        if (text.isEmpty()) return intArrayOf(0)

        val result = mutableListOf(0)
        var codePoint = text.codePointAt(0)
        var cursor = Character.charCount(codePoint)
        var categoryBefore = Grapheme17Data.category(codePoint)
        var state = nextState(1, categoryBefore, codePoint)

        while (cursor < text.length) {
            codePoint = text.codePointAt(cursor)
            val categoryAfter = Grapheme17Data.category(codePoint)
            val pairMask = Grapheme17Data.pairMasks[(categoryBefore shl 4) or categoryAfter]
            if (state and pairMask == 0) result += cursor
            state = nextState(state, categoryAfter, codePoint)
            cursor += Character.charCount(codePoint)
            categoryBefore = categoryAfter
        }
        result += text.length
        return result.toIntArray()
    }

    private fun nextState(state: Int, category: Int, codePoint: Int): Int = when (category) {
        3 -> if (state and 32 != 0) nextExtend(state, codePoint) else state and 21
        4 -> 17
        10 -> (state and 2) xor 3
        14 -> ((state and 16) shr 2) or (state and 41)
        15 -> 33
        else -> 1
    }

    private fun nextExtend(state: Int, codePoint: Int): Int {
        if (codePoint == 0x200c) return state and 21
        if (state and 8 != 0 || Grapheme17Data.isLinker(codePoint)) {
            return (state and 21) or 40
        }
        return (state and 21) or 32
    }
}
